### PR TITLE
Added Keycloak Authorization configuration

### LINKED ISSUE #2428: Integrate OAuth Authorization into Strimzi
This issue is about integrating into Strimzi the OAuth Authorization provided by Keycloak.
It's related to this https://github.com/strimzi/strimzi-kafka-oauth/pull/24 and providing the right CRD additions into the `Kafka` resource for enabling this type of authorization (other than the `simple` one already in place).
Authorization should have a new `keycloak` type and the related configuration parameters in the CRD.
Following a proposal for that.

```yaml
# ...
authorization:
  type: keycloak
  tokenEndpointUri: <URL of the Keycloak's token endpoint>
  clientId: <A client id of the OAuth client definition in Keycloak>
  delegateToKafkaAcls: false #if fallback to Simple authz if no access provided by Keycloak
  tlsTrustedCertificates:
     - secretName: oauth-server-cert
       certificate: ca.crt
  disableTlsHostnameVerification: false
  superUsers:
    - Alice
    - CN=Bob
```

### PR DESCRIPTION
**Pull request by ppatierno**
Signed-off-by: Paolo Patierno <ppatierno@live.com>

### Type of change

- Enhancement / new feature

### Description

This PR fixes #2428.
It's still missing tests and including the `kafka-oauth-keycloak-authorizer` JAR coming from this other PR https://github.com/strimzi/strimzi-kafka-oauth/pull/24 still in progress.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by ppatierno** [comment]
@scholzj @mstruk even if WIP, a first review would be really appreciated to have it ready as soon as possible for the next 0.17.0 release.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java [hunk 1]
```diff
@@ -369,32 +370,64 @@ private String getSecurityProtocol(boolean tls, boolean sasl)   {
...
+            writer.println("authorizer.class.name=" + KafkaAuthorizationSimple.AUTHORIZER_CLASS_NAME);
+
+            // User configured super users
+            if (simpleAuthz.getSuperUsers() != null && simpleAuthz.getSuperUsers().size() > 0) {
+                superUsers.addAll(simpleAuthz.getSuperUsers().stream().map(e -> String.format("User:%s", e)).collect(Collectors.toList()));
+            }
+        } else if (KafkaAuthorizationKeycloak.TYPE_KEYCLOAK.equals(authorization.getType())) {
+            KafkaAuthorizationKeycloak keycloakAuthz = (KafkaAuthorizationKeycloak) authorization;
+            writer.println("authorizer.class.name=" + KafkaAuthorizationKeycloak.AUTHORIZER_CLASS_NAME);
+            writer.println("principal.builder.class=" + KafkaAuthorizationKeycloak.PRINCIPAL_BUILDER_CLASS_NAME);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java]: Can you double check what impact does this have on the internal super users? Do they still keey their own names?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java [hunk 2]
```diff
@@ -369,32 +370,64 @@ private String getSecurityProtocol(boolean tls, boolean sasl)   {
...
     }
 
+    /**
+     * Configures authorization for the Kafka brokers. This method is used only internally.
+     *
+     * @param clusterName Name of the cluster
+     * @param superUsers Super users list who have all the rights on the cluster
+     * @param authorization The authorization configuration from the Kafka CR
+     */
+    private void configureAuthorization(String clusterName, List<String> superUsers, KafkaAuthorization authorization) {
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java]: We already have a similar pattern for configuring authentication on which I agree. The `configuraAuthentication` method is just one and we have the logic for different authentication mechanisms inside. I am not sure about the advantage of having different methods here.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 3]
```diff
@@ -1475,6 +1484,11 @@ private void setDataVolumesClaimsAndMountPaths(Storage storage) {
             }
         }
 
+        if (authorization instanceof KafkaAuthorizationKeycloak) {
+            KafkaAuthorizationKeycloak keycloakAuthz = (KafkaAuthorizationKeycloak) authorization;
+            volumeMountList.addAll(AuthenticationUtils.configureOauthCertificateVolumeMounts("authz-keycloak", keycloakAuthz.getTlsTrustedCertificates(), OAUTH_TRUSTED_CERTS_BASE_VOLUME_MOUNT + "/authz-keycloak-certs"));
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: yep, I forgot it :-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java (hunk 1 above)
Comment (mstruk) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java]: They should keep their names, yes.

The JwtKafkaPrincipalBuilder only does something different when user was authenticated over SASL_OAUTHBEARER as seen [here](https://github.com/mstruk/strimzi-kafka-oauth/blob/authz/oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/JwtKafkaPrincipalBuilder.java). An internal user would be handled by the same logic as before.

**Comment by ppatierno** [comment]
This is not WIP anymore but we don't have to merge until https://github.com/strimzi/strimzi-kafka-oauth/pull/24 is merged. Anyway can you have another review pass @scholzj @mstruk please? Thanks!

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public List<CertSecretSource> getTlsTrustedCertificates() {
+        return tlsTrustedCertificates;
+    }
+
+    public void setTlsTrustedCertificates(List<CertSecretSource> tlsTrustedCertificates) {
+        this.tlsTrustedCertificates = tlsTrustedCertificates;
+    }
+
+    @Description("Enable or disable TLS hostname verification. " +
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java]: hostname verification of _what_? Presumably the authz serverm but you should say that.

And should it be "enable...", defaulting to true?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java [hunk 5]
```diff
@@ -3222,4 +3223,83 @@ public void testCustomCertificateTls() {
...
+        KafkaCluster kc = KafkaCluster.fromCrd(kafkaAssembly, VERSIONS);
+        StatefulSet sts = kc.generateStatefulSet(true, null, null);
+        Container cont = sts.getSpec().getTemplate().getSpec().getContainers().get(0);
+
+        // Volume mounts
+        assertThat(cont.getVolumeMounts().stream().filter(mount -> "authz-keycloak-0".equals(mount.getName())).findFirst().orElse(null).getMountPath(), is(KafkaCluster.OAUTH_TRUSTED_CERTS_BASE_VOLUME_MOUNT + "/authz-keycloak-certs/first-certificate-0"));
+        assertThat(cont.getVolumeMounts().stream().filter(mount -> "authz-keycloak-1".equals(mount.getName())).findFirst().orElse(null).getMountPath(), is(KafkaCluster.OAUTH_TRUSTED_CERTS_BASE_VOLUME_MOUNT + "/authz-keycloak-certs/second-certificate-1"));
+
+        // Volumes
+        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "authz-keycloak-0".equals(vol.getName())).findFirst().orElse(null).getSecret().getItems().size(), is(1));
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java]: There are a couple of variables you can usefully factor out here.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java (hunk 4 above)
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java]: I used the same convention of the other OAuth authentication classes (i.e. see `KafkaListenerAuthenticationOAuth`)

Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java]: OK

**Comment by ppatierno** [comment]
@tombentley @scholzj @mstruk FYI I tried this with the keycloak library 0.3.0 (from the staging repository) and the integration seems to work fine.

**Comment by ppatierno** [comment]
@strimzi-ci run tests

**Comment by ppatierno** [comment]
@strimzi-ci run tests

File: docker-images/kafka/kafka-thirdparty-libs/2.3.x/pom.xml [hunk 6]
```diff
@@ -107,5 +107,35 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>io.strimzi</groupId>
+            <artifactId>kafka-oauth-keycloak-authorizer</artifactId>
+            <version>${strimzi-oauth-callback.version}</version>
```
Comment (scholzj) [inline docker-images/kafka/kafka-thirdparty-libs/2.3.x/pom.xml]: Should we change the name of the property? It is not just cllback anymore, so maybe `strimzi-oauth.version` would be better

File: docker-images/kafka/kafka-thirdparty-libs/2.3.x/pom.xml [hunk 7]
```diff
@@ -107,5 +107,35 @@
...
+                    <artifactId>jackson-core</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
     </dependencies>
+
+    <repositories>
+        <repository>
+            <id>staging</id>
+            <url>https://oss.sonatype.org/content/repositories/iostrimzi-1054/</url>
```
Comment (scholzj) [inline docker-images/kafka/kafka-thirdparty-libs/2.3.x/pom.xml]: We are now released, so we can remov this again.

**Pull request merged**